xla/service/gpu/gpu_compiler.cc

```diff
@@ -2003,34 +2010,45 @@ absl::StatusOr<GpuCompiler::BackendCompileResult> GpuCompiler::CompileAndLink(
         ++loaded_kernel_count;
       }
-      VLOG(2) << "Loaded " << loaded_kernel_count << " / "
-              << cache.entries_size() << " cached kernels.";
-    } else {
-      auto entries = cache.mutable_entries();
-      for (const auto& [name, binary] : binaries_to_cache) {
-        auto it = entries->find(name);
-        if (it == entries->end()) {
-          continue;
+      VLOG(2) << "Using " << loaded_kernel_count << " / "
+              << current_cache.entries_size() << " cached kernels.";
+    }
+    if (!binaries_to_cache.empty()) {
+      CompilationCacheProto disk_cache;
+      if (cache_file_exists) {
+        // Load the unmodified existing cache file to augment it.
+        std::string serialized;
+        TF_RETURN_IF_ERROR(tsl::ReadFileToString(tsl::Env::Default(),
+                                                 resolved_path, &serialized));
+        if (!disk_cache.ParseFromString(std::string(serialized))) {
+          return Internal("Failed to parse serialized CompilationCacheProto.");
         }
-        it->second.set_binary(reinterpret_cast<const char*>(binary.data()),
-                              binary.size());
-        VLOG(5) << "Cached kernels: " << name << ": " << binary.size();
       }
-      for (auto it = entries->begin(); it != entries->end();) {
-        if (it->second.binary().empty()) {
-          it = entries->erase(it);
-        } else {
-          ++it;
-        }
+      // Add kernels from the current compilation to the cache file. Binaries
+      // are taken from binaries_to_cache, all other kernel properties are
+      // copied from current_cache.
+      auto entries = disk_cache.mutable_entries();
+      int stored_kernel_count = 0;
+      for (const auto& [name, binary] : binaries_to_cache) {
+        auto it_current = current_cache.entries().find(name);
+        TF_RET_CHECK(it_current != current_cache.entries().end());
+        auto [it_disk, inserted] = entries->insert({name, it_current->second});
+        TF_RET_CHECK(inserted);
+        TF_RET_CHECK(!binary.empty());
+        it_disk->second.set_binary(reinterpret_cast<const char*>(binary.data()),
+                                   binary.size());
+        VLOG(5) << "Cached kernel: " << name << ": " << binary.size();
+        ++stored_kernel_count;
       }
-      if (cache.entries_size() > 0) {
-        TF_RETURN_IF_ERROR(tsl::WriteStringToFile(
-            tsl::Env::Default(), resolved_path, cache.SerializeAsString()));
-        VLOG(2) << "Stored " << cache.entries_size() << " / "
-                << binaries_to_cache.size();
+      if (stored_kernel_count > 0) {
+        TF_RETURN_IF_ERROR(
+            tsl::WriteStringToFile(tsl::Env::Default(), resolved_path,
+                                   disk_cache.SerializeAsString()));
+        VLOG(2) << "Stored " << stored_kernel_count << " / "
+                << binaries_to_cache.size() << " kernels in the cache file.";
       }
     }
```

Comment:
Maybe this could go to 1 or 2 separate function(s). We could also consider this for the lines 1985-2017, but I can't add a comment there.

Reply:
Moved this part into UpdateDiskKernelCache(). The other part does not look clearer as a separate function - I tried.